1-Draft/RFC0000-Chain-Operators.md

```diff
@@ -0,0 +1,611 @@
+pipeline background operators (`&`)
+or statement separators (`;`).
+
+For example:
+
+```none
+cmd1 && cmd2 &
+```
+
+Will bind as:
+
+```none
+[[cmd1 && cmd2] &]
+```
+
+Having the syntax tree:
+
+```none
+            <statement>
+           /           \
+         "&&"          "&"
+        /    \
+     "cmd1"  "cmd2"
+```
+
+The consequence of this will be that an entire pipeline chain
+can be sent to a background job for evaluation,
+rather than individual pipelines within it.
+
+### Semantics
+
+#### Pipeline "success"
+
+The `&&` and `||` operators proceed based on the "success"
+of the previous pipeline.
+
+The marker of command success proposed is the `$?` automatic variable.
+This is proposed because:
+
+- It builds off an existing PowerShell concept
+- It applies to both native commands and cmdlets/functions
+- Diverging from this would create inconsistency
+- Changes to the behaviour of `$?` would be reasonably expected to change here
+
+That is, there are the following direct equivalents:
+
+```none
+cmd1 && cmd2
+
+    |
+    v
+
+$(cmd1; if ($?) { cmd2 })
+```
+
+```none
+cmd1 || cmd2
+
+    |
+    v
+
+$(cmd1; if (-not $?) { cmd2 })
+```
+
+```none
+cmd1 && cmd2 || cmd3
+
+    |
+    v
+
+$(cmd1; if ($?) { cmd2 }; if (-not $?) { cmd3 })
+# Note that cmd1 failing runs cmd3
+```
+
+Also see the [alternate proposals section](#alternate-proposals-and-considerations).
+
+#### Pipeline output
+
+The output of a pipeline chain is the concatenation of all its pipelines,
+in the sequence of their output:
+
+```powershell
+$x = 'a','b','c' | Write-Output && Write-Output 'd'
+$x # 'a','b','c','d'
+```
+
+```powershell
+$x = 'a','b','c' | Write-Output && Write-Error 'Bad' || Write-Output 'd'
+# Writes the error record 'Bad'
+$x # 'a','b','c','d'
+```
+
+Commands that fail will have any pipeline output
+emitted before evaluation of the chain operator.
+
+For example:
+
+```powershell
+$x = cmd1 && cmd2
+```
+
+If `cmd1` fails but emits output, `$x` will hold that value.
+
+#### Error semantics
+
+Errors will have the same semantics as the equivalent
+`cmd1; if ($?) { cmd2 }` syntax.
+
+Terminating errors will terminate the entire pipeline chain.
+But output already emitted by earlier pipelines in the chain
+will be output by the chain before terminating:
+
+```powershell
+try
+{
+    $x = cmd1 && $(throw "Bad")
+}
+catch
+{
+}
+
+$x # Has values from cmd1
+```
+
+Non-terminating errors will cause the immediate pipeline to continue,
+and the pipeline chain will evaluate as normal based on the value of `$?`.
+
+### Other notes
+
+#### New Abstract Syntax Tree (AST) types
+
+The new AST type, `PipelineChainAst`,
+will be created to represent a pipeline chain.
+This will inherit from `PipelineBaseAst`
+and may occur anywhere a `PipelineBaseAst` does in a PowerShell AST.
+
+`ICustomAstVisitor2` and `AstVisitor2` would be extended to deal with this AST,
+and .NET Core 3's new default interface implementation feature would be

```

Comment:
Will the default interface implementation should throw an exception or do nothing?
I think doing nothing would be consistent with the old way of introducing a new interface when changing the Ast, but it also hides the fact that implementors might be missing something they should be implementing.

Reply:
@lzybkr to that point I think it would be a good time to introduce a `DefaultVisit`, that way in the future the implementer can choose how to react to a new AST type.

That doesn't really help for this round of changes though. Personally I would prefer that it do nothing.  Most of my `ICustomAstVisitor2`'s will continue to work *for the most part*.  And at the very least PowerShellEditorServices would have some hoops to jump through to get a significantly amount of analysis visitors working again otherwise.

Reply:
`DefaultVisit` is a great idea.

Reply:
Any of you mind elaborating the `DefaultVisit` idea? 😄

Reply:
@daxian-dbw basically this:

```csharp
interface ICustomAstVisitor
{
    object DefaultVisit(Ast ast) => null;

    object VisitCommandAst(CommandAst commandAst) => DefaultVisit(commandAst);

    object VisitNewAst(NewAst newAst) => DefaultVisit(newAst);
}
```

That way if you haven't implemented `VisitNewAst`, but you have implemented `DefaultVisit`, you can choose how to react to new AST types.  [roslyn](https://github.com/dotnet/roslyn/blob/6ead316e86d304e8b5bf2b85b71cbdf35bb2b52c/src/Compilers/CSharp/Portable/Syntax/CSharpSyntaxVisitor.cs#L31-L34) does this with it's syntax visitor.